File: app/src/main/java/net/bloople/manga/ReadingActivity.java
package net.bloople.manga;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager2.widget.ViewPager2;

public class ReadingActivity extends AppCompatActivity {
    private ReadingSession session;
    private ViewPager2 pager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_reading);

        pager = findViewById(R.id.pager);

        ImageView prev10 = findViewById(R.id.prev_10);
        prev10.setOnClickListener(ThrottledOnClickListener.wrap(v -> session.go(-10)));

        ImageView next10 = findViewById(R.id.next_10);
        next10.setOnClickListener(ThrottledOnClickListener.wrap(v -> session.go(10)));

        final Intent intent = getIntent();
        long libraryId = intent.getLongExtra("libraryId", -1);

        LibraryService.ensureLibrary(this, libraryId, library -> {
            if(library == null) return;
            long bookId = intent.getLongExtra("_id", -1);

            session = new ReadingSession(getApplicationContext(), library, library.getBooks().get(bookId));
            session.bind(this, pager);

            if(intent.getBooleanExtra("resume", false)) session.resume();

            if(savedInstanceState != null) {
                int pageFromBundle = savedInstanceState.getInt("page", -1);
                if(pageFromBundle != -1) session.page(pageFromBundle);
            }

            session.start();
        });
    }

    @Override
    public void onRestart() {
        super.onRestart();
        session.start();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putInt("page", session.page());
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onStop() {
        session.finish();
        super.onStop();
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfiguration) {
        super.onConfigurationChanged(newConfiguration);

        int currentItem = pager.getCurrentItem();
        pager.setAdapter(pager.getAdapter());
        pager.setCurrentItem(currentItem, false);
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemUI();
        }
    }

    private void hideSystemUI() {
        // Enables "sticky immersive" mode
        View decorView = getWindow().getDecorView();
        decorView.setSystemUiVisibility(
            View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                // Set the content to appear under the system bars so that the
                // content doesn't resize when the system bars hide and show.
                | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                // Hide the nav bar and status bar
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_FULLSCREEN);
    }
}

File: app/src/main/java/net/bloople/manga/ReadingActivity.kt
package net.bloople.manga

import android.content.res.Configuration
import android.os.Bundle
import android.widget.ImageView
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.widget.ViewPager2
import kotlinx.coroutines.launch
import net.bloople.manga.ThrottledOnClickListener.Companion.wrap


class ReadingActivity : AppCompatActivity() {
    private lateinit var session: ReadingSession
    private lateinit var pager: ViewPager2

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_reading)

        val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
        windowInsetsController.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())

        pager = findViewById(R.id.pager)

        val prev10 = findViewById<ImageView>(R.id.prev_10)
        prev10.setOnClickListener(wrap { session.go(-10) })

        val next10 = findViewById<ImageView>(R.id.next_10)
        next10.setOnClickListener(wrap { session.go(10) })

        val intent = intent

        val libraryId = intent.getLongExtra("libraryId", -1)

        lifecycleScope.launch {
            val library = LibraryService.ensureLibrary(this@ReadingActivity, libraryId) ?: return@launch

            val bookId = intent.getLongExtra("_id", -1)
            val book = library.books[bookId] ?: return@launch

            session = ReadingSession(library, book)
            session.bind(this@ReadingActivity, pager)

            if(intent.getBooleanExtra("resume", false)) session.resume()
            if(savedInstanceState != null) {
                val pageFromBundle = savedInstanceState.getInt("page", -1)
                if(pageFromBundle != -1) session.page(pageFromBundle)
            }

            session.start()
        }
    }

    public override fun onRestart() {
        super.onRestart()
        if(::session.isInitialized) session.start()
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt("page", session.page())
        super.onSaveInstanceState(outState)
    }

    public override fun onStop() {
        if(::session.isInitialized) session.finish()
        super.onStop()
    }

    override fun onConfigurationChanged(newConfiguration: Configuration) {
        super.onConfigurationChanged(newConfiguration)
        val currentItem = pager.currentItem
        pager.adapter = pager.adapter
        pager.setCurrentItem(currentItem, false)
    }
}